<?xml version="1.0" encoding="utf-8"?>
<!-- This URDF was automatically created by SolidWorks to URDF Exporter! Originally created by
Stephen Brawner (brawner@gmail.com) 
     Commit Version: 1.6.0-4-g7f85cfe  Build Version: 1.6.7995.38578
     For more information, please see http://wiki.ros.org/sw_urdf_exporter -->
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="tracked_robot_2">

	<xacro:include filename="inertial_macros.xacro" />

	<xacro:property name="mass" value="15"> </xacro:property>
    <xacro:property name="base_mass" value="15" />
    <xacro:property name="track_mass" value="10" />
    <xacro:property name="flipper_mass" value="2.5" />
    <xacro:property name="flipper_length" value="0.139577" />

	<!-- BASE LINK -->
	<link name="base_link">

	</link>

	<joint name="chassis_joint" type="fixed">
		<parent link="base_link" />
		<child link="chassis_link" />
		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	</joint>

	<!-- CHASSIS LINK -->
	<link name="chassis_link">

		<!-- <xacro:inertial_box mass="${mass}" x="0.25932" y="0.153" z="0.05932">
			<origin xyz="1.5315E-06 -1.1582E-06 0.024757" rpy="0 0 0" />
		</xacro:inertial_box> -->

        <inertial>
			<origin xyz="0 0 0.024757" rpy="0 0 0" />
			<mass
			value="15.0" />
		  <inertia
			ixx="5"
			ixy="0"
			ixz="0"
			iyy="5"
			iyz="0"
			izz="5" />
        </inertial>

		<visual>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/chassis_link.STL" />
			</geometry>
		</visual>

		<collision>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/chassis_link.STL" />
			</geometry>

		</collision>

	</link>

	<gazebo reference="chassis_link">
		<mu1>10</mu1>
		<mu2>10</mu2>
		<self_collide>1</self_collide>
		<kp value="10000000000.0" />
		<kd value="10.0" />
	</gazebo>

	<link name="imu_link">

		<!-- <xacro:inertial_box mass="${mass}" x="0.25932" y="0.153" z="0.05932">
			<origin xyz="1.5315E-06 -1.1582E-06 0.024757" rpy="0 0 0" />
		</xacro:inertial_box> -->

        <inertial>
			<origin xyz="0 0 0.024757" rpy="0 0 0" />
			<mass
			value="15.0" />
		  <inertia
			ixx="5"
			ixy="0"
			ixz="0"
			iyy="5"
			iyz="0"
			izz="5" />
        </inertial>

		<visual>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/chassis_link.STL" />
			</geometry>
		</visual>

		<collision>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/chassis_link.STL" />
			</geometry>

		</collision>

	</link>

	<gazebo reference="imu_link">
		<mu1>10</mu1>
		<mu2>10</mu2>
		<self_collide>1</self_collide>
		<kp value="10000000000.0" />
		<kd value="10.0" />

			<gravity>true</gravity>
			<sensor name="imu_sensor" type="imu">
			  <always_on>true</always_on>
			  <update_rate>100</update_rate>
			  <visualize>true</visualize>
			  <topic>__default_topic__</topic>
			  <plugin filename="libgazebo_ros_imu_sensor.so" name="imu_plugin">
				<topicName>imu</topicName>
				<bodyName>imu_link</bodyName>
				<updateRateHZ>10.0</updateRateHZ>
				<gaussianNoise>0.0</gaussianNoise>
				<xyzOffset>0 0 0</xyzOffset>
				<rpyOffset>0 0 0</rpyOffset>
				<frameName>imu_link</frameName>
				<initialOrientationAsReference>false</initialOrientationAsReference>
			  </plugin>
			  <pose>0 0 0 0 0 0</pose>
			</sensor>
	</gazebo>

	<joint name="imu_joint" type="fixed">
		<parent link="chassis_link" />
		<child link="imu_link" />
		<origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
	</joint>
	<!-- LEFT FLIPPER LINK -->
	<link name="left_flipper_link">
<!-- 
		<xacro:inertial_box mass="${flipper_mass}" x="0.139577" y="0.045" z="0.06958">
			<origin xyz="0.033943 0.00023886 0.011429" rpy="0 0 0" />
		</xacro:inertial_box> -->

        <inertial>
            <origin xyz="0.034059 0.0 -0.0" rpy="0 0 0" />
            <mass
        value="5" />
		<inertia
		ixx="5"
		ixy="0"
		ixz="0"
		iyy="5"
		iyz="0"
		izz="5" />
        </inertial>

		<visual>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/left_flipper_link.STL" />
			</geometry>

		</visual>

		<collision>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/left_flipper_link.STL" />
			</geometry>

		</collision>
	</link>

	<joint name="left_flipper_joint" type="continuous">
		<origin xyz="0.099884 -0.1132 0.029559" rpy="1.5708 0 0" />
		<parent link="base_link" />
		<child link="left_flipper_link" />
		<axis xyz="0 0 1" />
		<limit lower="-3" upper="3" effort="0.5" velocity="1.0" />
	</joint>

	<gazebo reference="left_flipper_link">
		<mu1>10</mu1>
		<mu2>10</mu2>
		<self_collide>1</self_collide>
		<material>Gazebo/DarkGrey</material>
		<kp value="10000000000.0" />
		<kd value="10.0" />
	</gazebo>

	<!-- LEFT FLIPPER LINK -->
	<link name="right_flipper_link">

		<!-- <xacro:inertial_box mass="${flipper_mass}" x="0.139577" y="0.045" z="0.06958">
			<origin xyz="0.034069 7.9651E-05 -0.00043099" rpy="0 0 0" />
		</xacro:inertial_box> -->

        <inertial>
            <origin xyz="0.034069 0 -0.0" rpy="0 0 0" />
            <mass
        value="5" />
		<inertia
		ixx="5"
		ixy="0"
		ixz="0"
		iyy="5"
		iyz="0"
		izz="5" />
        </inertial>

		<visual>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/right_flipper_link.STL" />
			</geometry>

		</visual>

		<collision>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/right_flipper_link.STL" />
			</geometry>

		</collision>
	</link>

	<joint name="right_flipper_joint" type="continuous">
		<origin xyz="0.099884 0.1132 0.029559" rpy="-1.5708 0 0" />
		<parent link="base_link" />
		<child link="right_flipper_link" />
		<axis xyz="0 0 1" />
		<limit lower="-3" upper="3" effort="0.5" velocity="1.0" />
	</joint>

	<gazebo reference="right_flipper_link">
		<mu1>10</mu1>
		<mu2>10</mu2>
		<self_collide>1</self_collide>
		<material>Gazebo/DarkGrey</material>
		<kp value="10000000000.0" />
		<kd value="10.0" />
	</gazebo>


	<!-- LEFT FLIPPER LINK -->
	<link name="rear_left_flipper_link">

		<!-- <xacro:inertial_box mass="${flipper_mass}" x="0.139577" y="0.045" z="0.06958">

			<origin xyz="-0.033863 -3.5938E-05 -0.00043099" rpy="0 0 0" />
		</xacro:inertial_box> -->

        <inertial>
            <origin xyz="-0.033863 0 0" rpy="0 0 0" />
            <mass
        value="5" />
		<inertia
		ixx="5"
		ixy="0"
		ixz="0"
		iyy="5"
		iyz="0"
		izz="5" />
        </inertial>

		<visual>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/rear_left_flipper_link.STL" />
			</geometry>

		</visual>

		<collision>

			<origin xyz="0 0 0" rpy="0 0 0" />
			<geometry>
				<mesh filename="$(find tracked_robot_2)/description/meshes/rear_left_flipper_link.STL" />
			</geometry>

		</collision>
	</link>

	<joint name="rear_left_flipper_joint" type="continuous">
		<origin xyz="-0.099884 -0.1132 0.029559" rpy="1.5708 0 0" />
		<parent link="base_link" />
		<child link="rear_left_flipper_link" />
		<axis xyz="0 0 1" />
		<limit lower="-3" upper="3" effort="0.5" velocity="1.0" />
	</joint>

	<gazebo reference="rear_left_flipper_link">
		<mu1>10</mu1>
		<mu2>10</mu2>
		<self_collide>1</self_collide>
		<material>Gazebo/DarkGrey</material>
		<kp value="10000000000.0" />
		<kd value="10.0" />
	</gazebo>

		<!-- LEFT FLIPPER LINK -->
		<link name="rear_right_flipper_link">

			<!-- <xacro:inertial_box mass="${flipper_mass}" x="0.139577" y="0.045" z="0.06958">

			<origin xyz="0.033863 7.9651E-05 -0.00043099" rpy="0 0 0" />
			</xacro:inertial_box>
	 -->
	 <inertial>
		<origin xyz="-0.034063 -0.0 -0.0" rpy="0 0 0" />
		<mass
        value="5" />
		<inertia
		ixx="5"
		ixy="0"
		ixz="0"
		iyy="5"
		iyz="0"
		izz="5" />
	</inertial>
			<visual>
	
				<origin xyz="0 0 0" rpy="0 0 0" />
				<geometry>
					<mesh filename="$(find tracked_robot_2)/description/meshes/rear_right_flipper_link.STL" />
				</geometry>
	
			</visual>
	
			<collision>
	
				<origin xyz="0 0 0" rpy="0 0 0" />
				<geometry>
					<mesh filename="$(find tracked_robot_2)/description/meshes/rear_right_flipper_link.STL" />
				</geometry>
	
			</collision>
		</link>
	
		<joint name="rear_right_flipper_joint" type="continuous">
			<origin xyz="-0.099884 0.1132 0.029559" rpy="-1.5708 0 0" />
			<parent link="base_link" />
			<child link="rear_right_flipper_link" />
			<axis xyz="0 0 1" />
			<limit lower="-3" upper="3" effort="0.5" velocity="1.0" />
		</joint>
	
		<gazebo reference="rear_right_flipper_link">
			<mu1>10</mu1>
			<mu2>10</mu2>
			<self_collide>1</self_collide>
			<material>Gazebo/DarkGrey</material>
			<kp value="1000000.0" />
			<kd value="10.0" />
		</gazebo>

	<gazebo reference="base_link">
		<self_collide>1</self_collide>
	</gazebo>

	<gazebo reference="chassis_link">
		<material>Gazebo/Grey</material>

		<self_collide>1</self_collide>
	</gazebo>

	<gazebo reference="left_flipper_link">
		<self_collide>1</self_collide>
	</gazebo>

	<gazebo reference="right_flipper_link">
		<self_collide>1</self_collide>
	</gazebo>

	<gazebo reference="rear_left_flipper_link">
		<self_collide>1</self_collide>
	</gazebo>

	<gazebo reference="rear_right_flipper_link">
		<self_collide>1</self_collide>
	</gazebo>


	<gazebo>
		<plugin filename='libgazebo_track_interface.so' name='tracked_vehicle_interface'>
			<commandROSTopic>/cmd_vel</commandROSTopic>
			<commandIGNTopic>~/tracked_robot_2/cmd_vel_twist</commandIGNTopic>
			<robotBaseFrame>base_link</robotBaseFrame>
			<tracks_separation>0.571</tracks_separation>
		</plugin>
	</gazebo>

	<gazebo>
		<plugin filename="libSimpleTrackedVehiclePlugin.so" name="hello">
			<!-- <robot_namespace>/MYROBOT</robot_namespace> -->
			<body>base_link</body>
			<left_track>left_flipper_link</left_track>
			<right_track>right_flipper_link</right_track>
			<left_flipper>rear_left_flipper_link</left_flipper>
			<right_flipper>rear_right_flipper_link</right_flipper>
			
			
			<!-- <left_flipper>rear_left_flipper_link</left_flipper>
			<right_flipper>rear_right_flipper_link</right_flipper> -->
	
			<track_mu>2</track_mu>
			<track_mu2>0.5</track_mu2>
			<max_angular_speed>0.1</max_angular_speed>
			<max_linear_speed>0.1</max_linear_speed>

		</plugin>
	</gazebo>


	<!-- ENSURE YOU CHECK THE GAZEBO PLUGIN PATH
	 echo $GAZEBO_PLUGIN_PATH

	 export GAZEBO_PLUGIN_PATH=/home/rob/ros_tut_ws/install/tracked_robot_2/share/tracked_robot_2
 -->

	<!-- <gazebo>
		<plugin filename="libhelloworld.so" name="hello"></plugin>
	</gazebo>
	 -->
	<gazebo>
		<plugin filename="libKeysToCmdVelPlugin.so" name="keyboard_control">
			<publish_as_twist>true</publish_as_twist>
			<!-- This MUST be the name of the ros package -->
			<cmd_vel_topic>~/tracked_robot_2/cmd_vel_twist</cmd_vel_topic>   
		</plugin>
	</gazebo>

	    <!-- <ros2_control name="GazeboSystem" type="system">
        <hardware>
            <plugin>gazebo_ros2_control/GazeboSystem</plugin>
        </hardware>
    <joint name="left_flipper_joint">
        <command_interface name="position">

        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity"/>
      </joint>

      <joint name="right_flipper_joint">
        <command_interface name="position">

        </command_interface>
        <state_interface name="position"/>
      </joint>

      <joint name="rear_left_flipper_joint">
        <command_interface name="position">

        </command_interface>
        <state_interface name="position"/>
      </joint>

      <joint name="rear_right_flipper_joint">
        <command_interface name="position">

        </command_interface>
        <state_interface name="position"/>
      </joint>

    </ros2_control>

    <gazebo>
        <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
            <parameters>$(find tracked_robot_2)/config/flipper_joint.yaml</parameters>
        </plugin>   
    </gazebo> -->

</robot>